File: dfs.py
from concurrent.futures import ThreadPoolExecutor


class Node:
    def __init__(self, data):
        self.data = data
        self.neighbors = []


def parallel_dfs(node, visited):

    if node in visited:
        return

    visited.add(node)
    print(node.data, end=" ")

    def work(neigh):
        if neigh not in visited:
            parallel_dfs(neigh, visited)

    with ThreadPoolExecutor() as executor:
        executor.map(work, node.neighbors)


# -------- Example Graph --------
a = Node(0)
b = Node(1)
c = Node(2)
d = Node(3)

a.neighbors = [b, c]
b.neighbors = [a, d]
c.neighbors = [a]
d.neighbors = [b]

print("\nParallel DFS:")
parallel_dfs(a, set())

File: dfs.cpp
#include <iostream>
#include <vector>
#include <omp.h>

using namespace std;

// ---------------- NODE STRUCTURE ----------------
struct Node {
    int data;
    vector<Node*> neighbors;
};

// ---------------- PARALLEL DFS ----------------
void parallelDFS(Node* node, vector<bool>& visited) {

    // check + mark visited safely
    bool isSafe = false;

    #pragma omp critical
    {
        if (!visited[node->data]) {
            visited[node->data] = true;
            cout << node->data << " ";
            isSafe = true;
        }
    }

    // if already visited, return
    if (!isSafe)
        return;

    // explore neighbors in parallel
    #pragma omp parallel for
    for (int i = 0; i < node->neighbors.size(); i++) {

        Node* neigh = node->neighbors[i];

        if (!visited[neigh->data]) {
            parallelDFS(neigh, visited);
        }
    }
}

// ---------------- MAIN FUNCTION ----------------
int main() {

    // Create graph (example)
    vector<Node> graph(5);

    for (int i = 0; i < 5; i++)
        graph[i].data = i;

    // Undirected graph
    graph[0].neighbors = {&graph[1], &graph[2]};
    graph[1].neighbors = {&graph[0], &graph[3]};
    graph[2].neighbors = {&graph[0], &graph[4]};
    graph[3].neighbors = {&graph[1]};
    graph[4].neighbors = {&graph[2]};

    vector<bool> visited(5, false);

    cout << "Parallel DFS Traversal: ";

    #pragma omp parallel
    {
        #pragma omp single
        parallelDFS(&graph[0], visited);
    }

    cout << endl;

    return 0;
}